File: Problems/875_kokoEatingBananas.ts
//[3,6,7,11], h = 8

//we have 8 hours to finish all bananas
//h must always be >= number of piles, therefore maximum k = maximum pile
//we want to see how many 'ticks' each pile requires for a k

//let p be any pile; the number of ticks this pile requires is ceil(p / k)
//start with k = max pile, e.g. 11

//ticks: [1, 1, 1, 1]; we want to increase the total number of ticks so that it equals h

//try to half k (choose lower bound): 5
//ticks: [1, 2, 2, 3], total ticks: 8
//current bounds: [1, 5]

//try to half again, k = 3
//ticks: [1, 2, 3, 4] -> now the total ticks > h, so this is invalid
//current bounds: [4, 5]

//k = 4
//ticks: [1, 2, 2, 3], total ticks: 8 -> valid
//current bounds [4, 4]

//break, return 4

//clear binary search pattern: we have a lower and upper bound for k: initially [1, max]
//continuously narrow search until upper bound == lower bound

function minEatingSpeed(piles: number[], h: number): number {
    const maxK = Math.max(...piles);
    let [lowerBound, upperBound] = [1, maxK];
    while (lowerBound < upperBound) {
        const midK = (lowerBound + upperBound) >> 1;
        const totalTicks = CountTicks(piles, midK);

        if (totalTicks <= h) { //we know the current k is valid, so bring upper bound down
            upperBound = midK
        }
        else { //current k is invalid, so bring lower bound to k + 1
            lowerBound = midK + 1;
        }
    }

    return lowerBound; //lowerBound = upperBound
};

const CountTicks = (piles: number[], k: number) => {
    let totalTicks = 0;
    for (const pile of piles) {
        const pileTicks = Math.ceil(pile / k)
        totalTicks += pileTicks;
    }
    return totalTicks;
}

console.log(minEatingSpeed([30,11,23,4,20], 6));

File: Problems/875_kokoEatingBananas.py
class Solution(object):
    def minEatingSpeed(self, piles, h):
        """
        :type piles: List[int]
        :type h: int
        :rtype: int
        """
        
        #we are looking for minimum integer k such that all bananas can be eaten in h hours, where h >= length of piles
        #the maximum value of k will be the maximum number of bananas in one pile (if we set k = max(piles) then each pile will take at most 1 tick to eat)
        #use binary search to hone in on a value for k, with lower bound 1

        #values of k: [1, 2, 3 ... max(piles)] can be mapped to their total times to eat all piles: [1000, 200, 30 ... len(piles)]
        #this list is sorted, and therefore we can use binary search to find the first item which is lower than or equal to h

        left = 1
        right = max(piles)

        while left < right:
            midK = (left + right) // 2

            #calculate total time taken to eat piles; time for each pile is given by math.ceil(piles[i] / k)
            totalTime = 0
            for pile in piles:
                totalTime -= ((pile * -1) // midK) #multiply by -1 the integer divide to get round-up rather than round down functionality

            if totalTime <= h: #we're in the correct portion of the list, but we could get a lower k value; move right to midK but still include midK
                right = midK 
            else: #totalTime > h; k is too small, so move left
                left = midK + 1

        #there is always a valid k in the list, so we can simply return k now (which is left or right; they are both the same at this point)
        return left

